File: backend/fx.py
"""FX + geolocation router.

Exposes:
    GET /api/fx/localize?base=NGN
        Detects the caller's country/currency from their IP, fetches a live
        FX rate from `base` -> detected currency, and returns everything in
        one round-trip. Result is cached in-process for 1 hour.

    GET /api/fx/rate?base=NGN&to=USD
        Returns just the FX rate between two currencies (cached 1h).

    GET /api/fx/currencies
        Curated list of currencies for the frontend selector dropdown.

No API keys required. Uses:
    - ipapi.co for IP geolocation
    - open.er-api.com for FX rates
"""

from __future__ import annotations

import logging
import time
from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Request

logger = logging.getLogger(__name__)

fx_router = APIRouter(prefix="/api/fx", tags=["fx"])

# ---------------------------------------------------------------------------
# In-process caches (simple TTL). For multi-instance deployments swap for Redis.
# ---------------------------------------------------------------------------

_RATE_TTL = 60 * 60  # 1 hour
_GEO_TTL = 24 * 60 * 60  # 24 hours

_rate_cache: dict[str, tuple[float, float]] = {}  # key "BASE:TARGET" -> (rate, exp_ts)
_all_rates_cache: dict[str, tuple[dict[str, float], float]] = {}  # base -> (rates, exp_ts)
_geo_cache: dict[str, tuple[dict[str, Any], float]] = {}  # ip -> (payload, exp_ts)


CURATED_CURRENCIES: list[dict[str, str]] = [
    {"code": "NGN", "name": "Nigerian Naira", "symbol": "\u20a6"},
    {"code": "USD", "name": "US Dollar", "symbol": "$"},
    {"code": "EUR", "name": "Euro", "symbol": "\u20ac"},
    {"code": "GBP", "name": "British Pound", "symbol": "\u00a3"},
    {"code": "CAD", "name": "Canadian Dollar", "symbol": "$"},
    {"code": "AUD", "name": "Australian Dollar", "symbol": "$"},
    {"code": "INR", "name": "Indian Rupee", "symbol": "\u20b9"},
    {"code": "ZAR", "name": "South African Rand", "symbol": "R"},
    {"code": "KES", "name": "Kenyan Shilling", "symbol": "KSh"},
    {"code": "GHS", "name": "Ghanaian Cedi", "symbol": "\u20b5"},
    {"code": "UGX", "name": "Ugandan Shilling", "symbol": "USh"},
    {"code": "TZS", "name": "Tanzanian Shilling", "symbol": "TSh"},
    {"code": "XOF", "name": "West African CFA Franc", "symbol": "CFA"},
    {"code": "XAF", "name": "Central African CFA Franc", "symbol": "FCFA"},
    {"code": "EGP", "name": "Egyptian Pound", "symbol": "\u00a3"},
    {"code": "MAD", "name": "Moroccan Dirham", "symbol": "DH"},
    {"code": "AED", "name": "UAE Dirham", "symbol": "AED"},
    {"code": "SAR", "name": "Saudi Riyal", "symbol": "SR"},
    {"code": "JPY", "name": "Japanese Yen", "symbol": "\u00a5"},
    {"code": "CNY", "name": "Chinese Yuan", "symbol": "\u00a5"},
    {"code": "SGD", "name": "Singapore Dollar", "symbol": "$"},
    {"code": "HKD", "name": "Hong Kong Dollar", "symbol": "$"},
    {"code": "CHF", "name": "Swiss Franc", "symbol": "CHF"},
    {"code": "SEK", "name": "Swedish Krona", "symbol": "kr"},
    {"code": "NOK", "name": "Norwegian Krone", "symbol": "kr"},
    {"code": "DKK", "name": "Danish Krone", "symbol": "kr"},
    {"code": "BRL", "name": "Brazilian Real", "symbol": "R$"},
    {"code": "MXN", "name": "Mexican Peso", "symbol": "$"},
    {"code": "TRY", "name": "Turkish Lira", "symbol": "\u20ba"},
]


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _client_ip(request: Request) -> str | None:
    """Best-effort caller IP extraction, honouring common proxy headers."""
    xff = request.headers.get("x-forwarded-for") or request.headers.get("x-real-ip")
    if xff:
        # X-Forwarded-For can be a comma-separated list; the leftmost is the client.
        ip = xff.split(",")[0].strip()
        if ip:
            return ip
    client = request.client
    return client.host if client else None


async def _fetch_all_rates(base: str) -> dict[str, float]:
    """Return a dict of {currency: rate} with `base` as the base. Cached 1h."""
    now = time.time()
    cached = _all_rates_cache.get(base)
    if cached and cached[1] > now:
        return cached[0]

    url = f"https://open.er-api.com/v6/latest/{base}"
    try:
        async with httpx.AsyncClient(timeout=6.0) as client:
            resp = await client.get(url)
            resp.raise_for_status()
            payload = resp.json()
    except (httpx.HTTPError, ValueError) as exc:
        logger.warning("FX rate fetch failed for base=%s: %s", base, exc)
        # Return whatever we last had (even if stale), otherwise raise.
        if cached:
            return cached[0]
        raise HTTPException(status_code=502, detail="Failed to fetch FX rates") from exc

    rates = payload.get("rates")
    if not isinstance(rates, dict) or not rates:
        if cached:
            return cached[0]
        raise HTTPException(status_code=502, detail="Malformed FX response")

    _all_rates_cache[base] = (rates, now + _RATE_TTL)
    return rates


async def _fetch_geo(ip: str) -> dict[str, Any]:
    """Fetch geolocation for an IP. Cached 24h.

    Uses ip-api.com (free, keyless, ~45 req/min per public IP). Falls back
    to an empty dict on any error so callers can degrade gracefully.
    """
    now = time.time()
    cached = _geo_cache.get(ip)
    if cached and cached[1] > now:
        return cached[0]

    url = (
        f"http://ip-api.com/json/{ip}"
        "?fields=status,country,countryCode,currency,languages"
    )
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.get(url)
            resp.raise_for_status()
            payload = resp.json()
    except (httpx.HTTPError, ValueError) as exc:
        logger.info("Geo lookup failed for ip=%s: %s", ip, exc)
        return {}

    if not isinstance(payload, dict) or payload.get("status") != "success":
        return {}

    result = {
        "country_code": payload.get("countryCode") or None,
        "country_name": payload.get("country") or None,
        "currency": (payload.get("currency") or "").upper() or None,
        "languages": payload.get("languages") or None,
    }
    _geo_cache[ip] = (result, now + _GEO_TTL)
    return result


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@fx_router.get("/currencies")
async def list_currencies() -> dict[str, Any]:
    """Curated currency list for the frontend selector."""
    return {"currencies": CURATED_CURRENCIES}


@fx_router.get("/rate")
async def get_rate(base: str = "NGN", to: str = "USD") -> dict[str, Any]:
    """Return the FX rate from `base` to `to`. Cached 1h."""
    base = base.upper().strip()
    to = to.upper().strip()

    if base == to:
        return {"base": base, "target": to, "rate": 1.0, "cached": False}

    key = f"{base}:{to}"
    now = time.time()
    cached = _rate_cache.get(key)
    if cached and cached[1] > now:
        return {"base": base, "target": to, "rate": cached[0], "cached": True}

    rates = await _fetch_all_rates(base)
    rate = rates.get(to)
    if not rate or not isinstance(rate, (int, float)) or rate <= 0:
        raise HTTPException(
            status_code=404,
            detail=f"No FX rate available for {base}->{to}",
        )

    _rate_cache[key] = (float(rate), now + _RATE_TTL)
    return {"base": base, "target": to, "rate": float(rate), "cached": False}


@fx_router.get("/localize")
async def localize(request: Request, base: str = "NGN") -> dict[str, Any]:
    """One-shot endpoint: detects caller's currency + returns FX rate."""
    base = base.upper().strip()

    ip = _client_ip(request)
    geo: dict[str, Any] = {}
    if ip and not _is_private_ip(ip):
        geo = await _fetch_geo(ip)

    detected_currency = (geo.get("currency") or base).upper()
    country_code = geo.get("country_code")
    languages = geo.get("languages") or ""
    locale = languages.split(",")[0].replace("_", "-") if languages else None

    # Same currency -> rate 1
    if detected_currency == base:
        return {
            "base": base,
            "detected_currency": base,
            "country_code": country_code,
            "locale": locale,
            "rate": 1.0,
            "source": "geo" if geo else "fallback",
        }

    rate: float = 1.0
    used_currency = detected_currency
    try:
        rates = await _fetch_all_rates(base)
        found = rates.get(detected_currency)
        if found and isinstance(found, (int, float)) and found > 0:
            rate = float(found)
        else:
            # No rate for detected currency -> fall back to base
            used_currency = base
            rate = 1.0
    except HTTPException:
        used_currency = base
        rate = 1.0

    return {
        "base": base,
        "detected_currency": used_currency,
        "country_code": country_code,
        "locale": locale,
        "rate": rate,
        "source": "geo" if geo else "fallback",
    }


def _is_private_ip(ip: str) -> bool:
    """Skip geolocation for LAN / loopback addresses (dev environments)."""
    try:
        import ipaddress

        addr = ipaddress.ip_address(ip)
        return addr.is_private or addr.is_loopback or addr.is_link_local
    except ValueError:
        return True

File: backend/routers/fx.js
"""
FX conversion and IP-based currency localization router.

Endpoints
---------
GET /api/fx/localize?base=NGN
    Detect the visitor's country and currency from their IP address, then
    return the FX rate from the base currency to the detected currency.

GET /api/fx/rate?base=NGN&to=USD
    Return the current FX rate between two supported currencies.

GET /api/fx/currencies
    Return the curated list of currencies supported by the frontend selector.

External services
-----------------
IP geolocation:
    https://ipapi.co

Foreign exchange rates:
    https://open.er-api.com

No API keys are required for these endpoints.

Important
---------
Localized prices should normally be treated as display estimates only.
The authoritative course price and payment amount should continue to come
from the database in the LMS base currency, such as NGN.
"""

from __future__ import annotations

import asyncio
import ipaddress
import logging
import re
import time
from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Query, Request, status

logger = logging.getLogger(__name__)

# Because this router already includes /api, mount it with:
#
#     app.include_router(fx_router)
#
# Do not add prefix="/api" again when mounting it.
fx_router = APIRouter(prefix="/api/fx", tags=["fx"])


# =============================================================================
# Configuration
# =============================================================================

RATE_TTL_SECONDS = 60 * 60
GEO_TTL_SECONDS = 24 * 60 * 60

FX_REQUEST_TIMEOUT = 8.0
GEO_REQUEST_TIMEOUT = 6.0

MAX_GEO_CACHE_ENTRIES = 5_000

DEFAULT_BASE_CURRENCY = "NGN"
DEFAULT_LOCALE = "en-NG"

CURRENCY_CODE_PATTERN = re.compile(r"^[A-Z]{3}$")


# =============================================================================
# Supported currencies
# =============================================================================

CURATED_CURRENCIES: list[dict[str, str]] = [
    {
        "code": "NGN",
        "name": "Nigerian Naira",
        "symbol": "₦",
        "default_locale": "en-NG",
    },
    {
        "code": "USD",
        "name": "US Dollar",
        "symbol": "$",
        "default_locale": "en-US",
    },
    {
        "code": "EUR",
        "name": "Euro",
        "symbol": "€",
        "default_locale": "en-IE",
    },
    {
        "code": "GBP",
        "name": "British Pound",
        "symbol": "£",
        "default_locale": "en-GB",
    },
    {
        "code": "CAD",
        "name": "Canadian Dollar",
        "symbol": "$",
        "default_locale": "en-CA",
    },
    {
        "code": "AUD",
        "name": "Australian Dollar",
        "symbol": "$",
        "default_locale": "en-AU",
    },
    {
        "code": "INR",
        "name": "Indian Rupee",
        "symbol": "₹",
        "default_locale": "en-IN",
    },
    {
        "code": "ZAR",
        "name": "South African Rand",
        "symbol": "R",
        "default_locale": "en-ZA",
    },
    {
        "code": "KES",
        "name": "Kenyan Shilling",
        "symbol": "KSh",
        "default_locale": "en-KE",
    },
    {
        "code": "GHS",
        "name": "Ghanaian Cedi",
        "symbol": "₵",
        "default_locale": "en-GH",
    },
    {
        "code": "UGX",
        "name": "Ugandan Shilling",
        "symbol": "USh",
        "default_locale": "en-UG",
    },
    {
        "code": "TZS",
        "name": "Tanzanian Shilling",
        "symbol": "TSh",
        "default_locale": "en-TZ",
    },
    {
        "code": "XOF",
        "name": "West African CFA Franc",
        "symbol": "CFA",
        "default_locale": "fr-SN",
    },
    {
        "code": "XAF",
        "name": "Central African CFA Franc",
        "symbol": "FCFA",
        "default_locale": "fr-CM",
    },
    {
        "code": "EGP",
        "name": "Egyptian Pound",
        "symbol": "E£",
        "default_locale": "ar-EG",
    },
    {
        "code": "MAD",
        "name": "Moroccan Dirham",
        "symbol": "DH",
        "default_locale": "ar-MA",
    },
    {
        "code": "AED",
        "name": "UAE Dirham",
        "symbol": "AED",
        "default_locale": "ar-AE",
    },
    {
        "code": "SAR",
        "name": "Saudi Riyal",
        "symbol": "SAR",
        "default_locale": "ar-SA",
    },
    {
        "code": "JPY",
        "name": "Japanese Yen",
        "symbol": "¥",
        "default_locale": "ja-JP",
    },
    {
        "code": "CNY",
        "name": "Chinese Yuan",
        "symbol": "¥",
        "default_locale": "zh-CN",
    },
    {
        "code": "SGD",
        "name": "Singapore Dollar",
        "symbol": "$",
        "default_locale": "en-SG",
    },
    {
        "code": "HKD",
        "name": "Hong Kong Dollar",
        "symbol": "$",
        "default_locale": "en-HK",
    },
    {
        "code": "CHF",
        "name": "Swiss Franc",
        "symbol": "CHF",
        "default_locale": "de-CH",
    },
    {
        "code": "SEK",
        "name": "Swedish Krona",
        "symbol": "kr",
        "default_locale": "sv-SE",
    },
    {
        "code": "NOK",
        "name": "Norwegian Krone",
        "symbol": "kr",
        "default_locale": "nb-NO",
    },
    {
        "code": "DKK",
        "name": "Danish Krone",
        "symbol": "kr",
        "default_locale": "da-DK",
    },
    {
        "code": "BRL",
        "name": "Brazilian Real",
        "symbol": "R$",
        "default_locale": "pt-BR",
    },
    {
        "code": "MXN",
        "name": "Mexican Peso",
        "symbol": "$",
        "default_locale": "es-MX",
    },
    {
        "code": "TRY",
        "name": "Turkish Lira",
        "symbol": "₺",
        "default_locale": "tr-TR",
    },
]

SUPPORTED_CURRENCIES = {
    currency["code"] for currency in CURATED_CURRENCIES
}

CURRENCY_METADATA = {
    currency["code"]: currency for currency in CURATED_CURRENCIES
}


# =============================================================================
# In-process caches
# =============================================================================

# "BASE:TARGET" -> (rate, expiry timestamp)
_rate_cache: dict[str, tuple[float, float]] = {}

# "BASE" -> (all rates, expiry timestamp)
_all_rates_cache: dict[str, tuple[dict[str, float], float]] = {}

# "IP" -> (geolocation result, expiry timestamp)
_geo_cache: dict[str, tuple[dict[str, Any], float]] = {}

# Prevent multiple simultaneous requests from fetching the same information.
_rates_lock = asyncio.Lock()
_geo_lock = asyncio.Lock()


# =============================================================================
# Validation and normalization helpers
# =============================================================================

def _normalize_currency(
    value: str,
    *,
    field_name: str,
    require_curated: bool = True,
) -> str:
    """
    Normalize and validate an ISO-style three-letter currency code.
    """
    code = value.strip().upper()

    if not CURRENCY_CODE_PATTERN.fullmatch(code):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid {field_name} currency code.",
        )

    if require_curated and code not in SUPPORTED_CURRENCIES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Unsupported {field_name} currency: {code}.",
        )

    return code


def _normalize_locale(value: str | None) -> str | None:
    """
    Convert locale values such as en_US to en-US.
    """
    if not value:
        return None

    first_locale = value.split(",")[0].strip()

    if not first_locale:
        return None

    return first_locale.replace("_", "-")


def _default_locale_for_currency(currency: str) -> str:
    """
    Return the configured default locale for a supported currency.
    """
    metadata = CURRENCY_METADATA.get(currency, {})

    return metadata.get("default_locale", DEFAULT_LOCALE)


def _currency_symbol(currency: str) -> str:
    """
    Return the configured display symbol for a supported currency.
    """
    metadata = CURRENCY_METADATA.get(currency, {})

    return metadata.get("symbol", currency)


# =============================================================================
# IP helpers
# =============================================================================

def _valid_ip(value: str | None) -> str | None:
    """
    Return a normalized IP address or None when invalid.
    """
    if not value:
        return None

    candidate = value.strip()

    try:
        return str(ipaddress.ip_address(candidate))
    except ValueError:
        return None


def _client_ip(request: Request) -> str | None:
    """
    Best-effort client IP extraction.

    Reverse proxies may add one or more of these headers. Every extracted
    value is validated before it is used.

    The order prioritizes common platform-controlled client-IP headers.
    """
    candidates: list[str | None] = [
        request.headers.get("cf-connecting-ip"),
        request.headers.get("true-client-ip"),
        request.headers.get("x-real-ip"),
    ]

    forwarded_for = request.headers.get("x-forwarded-for")

    if forwarded_for:
        # X-Forwarded-For may contain:
        # client, proxy-1, proxy-2
        #
        # The leftmost valid value is normally the original client.
        forwarded_candidates = [
            item.strip()
            for item in forwarded_for.split(",")
            if item.strip()
        ]

        candidates.extend(forwarded_candidates)

    if request.client:
        candidates.append(request.client.host)

    for candidate in candidates:
        valid = _valid_ip(candidate)

        if valid:
            return valid

    return None


def _is_non_public_ip(ip: str) -> bool:
    """
    Return True for private, loopback, reserved, multicast, link-local,
    or otherwise non-global addresses.
    """
    try:
        address = ipaddress.ip_address(ip)

        return not address.is_global
    except ValueError:
        return True


# =============================================================================
# Cache helpers
# =============================================================================

def _remove_expired_cache_entries() -> None:
    """
    Remove expired items from all in-process caches.
    """
    now = time.time()

    expired_rate_keys = [
        key
        for key, (_, expires_at) in _rate_cache.items()
        if expires_at <= now
    ]

    for key in expired_rate_keys:
        _rate_cache.pop(key, None)

    expired_all_rate_keys = [
        key
        for key, (_, expires_at) in _all_rates_cache.items()
        if expires_at <= now
    ]

    for key in expired_all_rate_keys:
        _all_rates_cache.pop(key, None)

    expired_geo_keys = [
        key
        for key, (_, expires_at) in _geo_cache.items()
        if expires_at <= now
    ]

    for key in expired_geo_keys:
        _geo_cache.pop(key, None)


def _trim_geo_cache() -> None:
    """
    Prevent the IP geolocation cache from growing indefinitely.
    """
    if len(_geo_cache) <= MAX_GEO_CACHE_ENTRIES:
        return

    sorted_entries = sorted(
        _geo_cache.items(),
        key=lambda item: item[1][1],
    )

    entries_to_remove = len(_geo_cache) - MAX_GEO_CACHE_ENTRIES

    for ip, _ in sorted_entries[:entries_to_remove]:
        _geo_cache.pop(ip, None)


# =============================================================================
# FX service
# =============================================================================

async def _fetch_all_rates(base: str) -> dict[str, float]:
    """
    Return all available rates for a base currency.

    Fresh values are cached for one hour. If the external service fails and
    an expired cached response exists, the expired response is used as a
    stale fallback.
    """
    now = time.time()
    cached = _all_rates_cache.get(base)

    if cached and cached[1] > now:
        return cached[0]

    async with _rates_lock:
        # Another request may have populated the cache while this request
        # waited for the lock.
        now = time.time()
        cached = _all_rates_cache.get(base)

        if cached and cached[1] > now:
            return cached[0]

        url = f"https://open.er-api.com/v6/latest/{base}"

        try:
            async with httpx.AsyncClient(
                timeout=FX_REQUEST_TIMEOUT,
                follow_redirects=True,
            ) as client:
                response = await client.get(
                    url,
                    headers={
                        "Accept": "application/json",
                        "User-Agent": "OrbalAcademy-LMS/1.0",
                    },
                )

                response.raise_for_status()
                payload = response.json()

        except (
            httpx.HTTPError,
            ValueError,
            TypeError,
        ) as exc:
            logger.warning(
                "FX rate request failed for base=%s: %s",
                base,
                exc,
            )

            if cached:
                logger.warning(
                    "Using stale FX cache for base=%s",
                    base,
                )

                return cached[0]

            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="The currency conversion service is temporarily unavailable.",
            ) from exc

        if not isinstance(payload, dict):
            if cached:
                return cached[0]

            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="The currency conversion service returned an invalid response.",
            )

        result_status = payload.get("result")

        if result_status and result_status != "success":
            logger.warning(
                "FX provider returned unsuccessful result for base=%s: %s",
                base,
                result_status,
            )

            if cached:
                return cached[0]

            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="The currency conversion service could not process the request.",
            )

        raw_rates = payload.get("rates")

        if not isinstance(raw_rates, dict) or not raw_rates:
            if cached:
                return cached[0]

            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="The currency conversion service returned no exchange rates.",
            )

        clean_rates: dict[str, float] = {}

        for currency_code, raw_rate in raw_rates.items():
            if not isinstance(currency_code, str):
                continue

            normalized_code = currency_code.strip().upper()

            if not CURRENCY_CODE_PATTERN.fullmatch(normalized_code):
                continue

            if isinstance(raw_rate, bool):
                continue

            if isinstance(raw_rate, (int, float)) and raw_rate > 0:
                clean_rates[normalized_code] = float(raw_rate)

        if not clean_rates:
            if cached:
                return cached[0]

            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="No valid exchange rates were returned.",
            )

        # Ensure the base currency always converts to itself.
        clean_rates[base] = 1.0

        _all_rates_cache[base] = (
            clean_rates,
            now + RATE_TTL_SECONDS,
        )

        return clean_rates


async def _get_exchange_rate(
    base: str,
    target: str,
) -> tuple[float, bool]:
    """
    Return an exchange rate and whether the individual rate cache was used.
    """
    if base == target:
        return 1.0, False

    cache_key = f"{base}:{target}"
    now = time.time()
    cached = _rate_cache.get(cache_key)

    if cached and cached[1] > now:
        return cached[0], True

    rates = await _fetch_all_rates(base)
    rate = rates.get(target)

    if (
        isinstance(rate, bool)
        or not isinstance(rate, (int, float))
        or rate <= 0
    ):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No FX rate is available for {base} to {target}.",
        )

    normalized_rate = float(rate)

    _rate_cache[cache_key] = (
        normalized_rate,
        now + RATE_TTL_SECONDS,
    )

    return normalized_rate, False


# =============================================================================
# Geolocation service
# =============================================================================

async def _fetch_geo(ip: str) -> dict[str, Any]:
    """
    Retrieve country, currency, and locale information for a public IP.

    Results are cached for 24 hours. Errors return an empty dictionary so
    localization can safely fall back to the LMS base currency.
    """
    now = time.time()
    cached = _geo_cache.get(ip)

    if cached and cached[1] > now:
        return cached[0]

    async with _geo_lock:
        now = time.time()
        cached = _geo_cache.get(ip)

        if cached and cached[1] > now:
            return cached[0]

        url = f"https://ipapi.co/{ip}/json/"

        try:
            async with httpx.AsyncClient(
                timeout=GEO_REQUEST_TIMEOUT,
                follow_redirects=True,
            ) as client:
                response = await client.get(
                    url,
                    headers={
                        "Accept": "application/json",
                        "User-Agent": "OrbalAcademy-LMS/1.0",
                    },
                )

                response.raise_for_status()
                payload = response.json()

        except (
            httpx.HTTPError,
            ValueError,
            TypeError,
        ) as exc:
            logger.info(
                "IP geolocation failed for ip=%s: %s",
                ip,
                exc,
            )

            if cached:
                return cached[0]

            return {}

        if not isinstance(payload, dict):
            return cached[0] if cached else {}

        if payload.get("error"):
            logger.info(
                "IP geolocation provider rejected ip=%s: %s",
                ip,
                payload.get("reason") or payload.get("message"),
            )

            return cached[0] if cached else {}

        country_code = payload.get("country_code")
        country_name = payload.get("country_name")
        currency = payload.get("currency")
        languages = payload.get("languages")

        normalized_country_code = (
            str(country_code).strip().upper()
            if country_code
            else None
        )

        normalized_currency = (
            str(currency).strip().upper()
            if currency
            else None
        )

        # A detected currency outside the curated list is retained in the
        # raw result, but /localize will safely fall back to the base currency.
        result: dict[str, Any] = {
            "country_code": normalized_country_code,
            "country_name": (
                str(country_name).strip()
                if country_name
                else None
            ),
            "currency": normalized_currency,
            "languages": (
                str(languages).strip()
                if languages
                else None
            ),
        }

        _geo_cache[ip] = (
            result,
            now + GEO_TTL_SECONDS,
        )

        _trim_geo_cache()

        return result


# =============================================================================
# API endpoints
# =============================================================================

@fx_router.get("/currencies")
async def list_currencies() -> dict[str, Any]:
    """
    Return currencies available in the frontend currency selector.
    """
    return {
        "base_currency": DEFAULT_BASE_CURRENCY,
        "currencies": CURATED_CURRENCIES,
        "count": len(CURATED_CURRENCIES),
    }


@fx_router.get("/rate")
async def get_rate(
    base: str = Query(
        default=DEFAULT_BASE_CURRENCY,
        min_length=3,
        max_length=3,
        description="The course-price base currency.",
    ),
    to: str = Query(
        default="USD",
        min_length=3,
        max_length=3,
        description="The target display currency.",
    ),
) -> dict[str, Any]:
    """
    Return the exchange rate from one supported currency to another.
    """
    normalized_base = _normalize_currency(
        base,
        field_name="base",
    )

    normalized_target = _normalize_currency(
        to,
        field_name="target",
    )

    rate, cached = await _get_exchange_rate(
        normalized_base,
        normalized_target,
    )

    _remove_expired_cache_entries()

    return {
        "base": normalized_base,
        "target": normalized_target,
        "rate": rate,
        "cached": cached,
        "base_symbol": _currency_symbol(normalized_base),
        "target_symbol": _currency_symbol(normalized_target),
        "locale": _default_locale_for_currency(normalized_target),
        "expires_in_seconds": RATE_TTL_SECONDS,
    }


@fx_router.get("/localize")
async def localize(
    request: Request,
    base: str = Query(
        default=DEFAULT_BASE_CURRENCY,
        min_length=3,
        max_length=3,
        description="The currency in which course prices are stored.",
    ),
) -> dict[str, Any]:
    """
    Detect the visitor's local currency and return its exchange rate.

    When geolocation or exchange-rate retrieval fails, the response falls
    back to the base currency with a rate of 1.0.
    """
    normalized_base = _normalize_currency(
        base,
        field_name="base",
    )

    client_ip = _client_ip(request)

    geo: dict[str, Any] = {}

    if client_ip and not _is_non_public_ip(client_ip):
        geo = await _fetch_geo(client_ip)

    detected_currency_raw = geo.get("currency")

    detected_currency = (
        str(detected_currency_raw).strip().upper()
        if detected_currency_raw
        else normalized_base
    )

    country_code = geo.get("country_code")
    country_name = geo.get("country_name")

    detected_locale = _normalize_locale(
        geo.get("languages")
    )

    # Use the detected currency only when it is in the curated list.
    if detected_currency not in SUPPORTED_CURRENCIES:
        logger.info(
            "Detected unsupported currency=%s; falling back to base=%s",
            detected_currency,
            normalized_base,
        )

        detected_currency = normalized_base

    locale = (
        detected_locale
        or _default_locale_for_currency(detected_currency)
    )

    source = "geo" if geo else "fallback"

    if detected_currency == normalized_base:
        _remove_expired_cache_entries()

        return {
            "base": normalized_base,
            "detected_currency": normalized_base,
            "country_code": country_code,
            "country_name": country_name,
            "locale": locale,
            "rate": 1.0,
            "source": source,
            "base_symbol": _currency_symbol(normalized_base),
            "currency_symbol": _currency_symbol(normalized_base),
            "is_converted": False,
            "is_estimate": False,
            "fallback_used": not bool(geo),
            "expires_in_seconds": RATE_TTL_SECONDS,
        }

    try:
        rate, _ = await _get_exchange_rate(
            normalized_base,
            detected_currency,
        )

        fallback_used = False
        used_currency = detected_currency

    except HTTPException as exc:
        logger.warning(
            "Could not localize currency from %s to %s: %s",
            normalized_base,
            detected_currency,
            exc.detail,
        )

        rate = 1.0
        used_currency = normalized_base
        locale = _default_locale_for_currency(normalized_base)
        source = "fallback"
        fallback_used = True

    _remove_expired_cache_entries()

    return {
        "base": normalized_base,
        "detected_currency": used_currency,
        "country_code": country_code,
        "country_name": country_name,
        "locale": locale,
        "rate": rate,
        "source": source,
        "base_symbol": _currency_symbol(normalized_base),
        "currency_symbol": _currency_symbol(used_currency),
        "is_converted": used_currency != normalized_base,
        "is_estimate": used_currency != normalized_base,
        "fallback_used": fallback_used,
        "expires_in_seconds": RATE_TTL_SECONDS,
    }
